File: cmd/kutup/internal/api/tus.go
// tus.io 1.0 protocol client.
//
// Companion to backend/handlers/tus.go. The desktop / mobile Tauri shells
// (eventually) and the CLI both speak this protocol to stream-upload big
// files without buffering in RAM. Resumability comes from the server-side
// Upload-Offset bookkeeping; this client just shuttles bytes.
package api

import (
	"bytes"
	"encoding/base64"
	"fmt"
	"io"
	"net/http"
	"strconv"
	"strings"
)

const tusVersion = "1.0.0"

// TusCreate opens a new tus session. `totalBytes` is the ciphertext byte
// count (the upload package's `CipherSize(plainBytes)` knows how to derive
// it). The five encrypted-metadata strings are the same base64 values the
// multipart /files/upload endpoint expects — we re-base64 them per tus
// spec so they round-trip cleanly through the Upload-Metadata header.
//
// Returns the upload ID extracted from the Location header.
func (c *Client) TusCreate(
	totalBytes int64,
	collectionID string,
	encryptedMetadata, metadataNonce string,
	encryptedFileKey, fileKeyNonce string,
) (string, error) {
	enc := base64.StdEncoding
	uploadMeta := strings.Join([]string{
		"collectionId " + enc.EncodeToString([]byte(collectionID)),
		"encryptedMetadata " + enc.EncodeToString([]byte(encryptedMetadata)),
		"metadataNonce " + enc.EncodeToString([]byte(metadataNonce)),
		"encryptedFileKey " + enc.EncodeToString([]byte(encryptedFileKey)),
		"fileKeyNonce " + enc.EncodeToString([]byte(fileKeyNonce)),
	}, ",")

	req, err := http.NewRequest(http.MethodPost, c.base+"/api/uploads/", nil)
	if err != nil {
		return "", err
	}
	req.Header.Set("Tus-Resumable", tusVersion)
	req.Header.Set("Upload-Length", strconv.FormatInt(totalBytes, 10))
	req.Header.Set("Upload-Metadata", uploadMeta)

	resp, err := c.do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusCreated {
		body, _ := io.ReadAll(resp.Body)
		return "", fmt.Errorf("tus create: HTTP %d: %s", resp.StatusCode, string(body))
	}

	loc := resp.Header.Get("Location")
	// Server returns absolute path like /api/uploads/<uuid>; take the
	// last path segment.
	if idx := strings.LastIndex(loc, "/"); idx >= 0 {
		return loc[idx+1:], nil
	}
	return "", fmt.Errorf("tus create: missing/garbled Location header %q", loc)
}

// TusHead returns the server's current Upload-Offset (the high-water mark
// of received bytes) and the original Upload-Length. Used for resume on
// the next CLI run; not exercised yet by v1's upload-from-scratch flow.
func (c *Client) TusHead(uploadID string) (offset, length int64, err error) {
	req, err := http.NewRequest(http.MethodHead, c.base+"/api/uploads/"+uploadID, nil)
	if err != nil {
		return 0, 0, err
	}
	req.Header.Set("Tus-Resumable", tusVersion)
	resp, err := c.do(req)
	if err != nil {
		return 0, 0, err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		return 0, 0, fmt.Errorf("tus head: HTTP %d", resp.StatusCode)
	}
	off, _ := strconv.ParseInt(resp.Header.Get("Upload-Offset"), 10, 64)
	ln, _ := strconv.ParseInt(resp.Header.Get("Upload-Length"), 10, 64)
	return off, ln, nil
}

// TusPatch ships one chunk. Returns the new offset reported by the server
// and the file ID (only set on the final chunk that triggers the
// finaliser; empty on intermediate chunks).
func (c *Client) TusPatch(uploadID string, offset int64, body []byte) (int64, string, error) {
	req, err := http.NewRequest(http.MethodPatch,
		c.base+"/api/uploads/"+uploadID,
		bytes.NewReader(body),
	)
	if err != nil {
		return 0, "", err
	}
	req.Header.Set("Tus-Resumable", tusVersion)
	req.Header.Set("Upload-Offset", strconv.FormatInt(offset, 10))
	req.Header.Set("Content-Type", "application/offset+octet-stream")
	req.Header.Set("Content-Length", strconv.Itoa(len(body)))
	// Each PATCH may be up to ~5 MB; bump beyond the client's default
	// 60 s so cold-storage SeaweedFS doesn't trip us on the first part.
	req.ContentLength = int64(len(body))

	resp, err := c.do(req)
	if err != nil {
		return 0, "", err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusNoContent {
		b, _ := io.ReadAll(resp.Body)
		return 0, "", fmt.Errorf("tus patch: HTTP %d: %s", resp.StatusCode, string(b))
	}
	newOff, _ := strconv.ParseInt(resp.Header.Get("Upload-Offset"), 10, 64)
	fileID := resp.Header.Get("X-Kutup-File-Id")
	return newOff, fileID, nil
}

// TusDelete cancels an in-flight upload. Best-effort — the server-side
// stale sweeper will reap abandoned sessions after 24h anyway.
func (c *Client) TusDelete(uploadID string) error {
	req, err := http.NewRequest(http.MethodDelete, c.base+"/api/uploads/"+uploadID, nil)
	if err != nil {
		return err
	}
	req.Header.Set("Tus-Resumable", tusVersion)
	resp, err := c.do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusNoContent && resp.StatusCode != http.StatusNotFound {
		return fmt.Errorf("tus delete: HTTP %d", resp.StatusCode)
	}
	return nil
}

File: crates/kutup-cli/src/api/tus.rs
//! tus.io 1.0 protocol client — mirrors `internal/api/tus.go`.
//! Companion to `backend/handlers/tus.go`.

use anyhow::{bail, Context, Result};
use base64::Engine;
use reqwest::Method;

use super::{api_error, Client};

const TUS_VERSION: &str = "1.0.0";

fn b64(bytes: &[u8]) -> String {
    base64::engine::general_purpose::STANDARD.encode(bytes)
}

impl Client {
    /// Opens a new tus session; `total_bytes` is the ciphertext byte count.
    /// Returns the upload id from the `Location` header. Mirrors `TusCreate`.
    #[allow(clippy::too_many_arguments)]
    pub fn tus_create(
        &self,
        total_bytes: i64,
        collection_id: &str,
        encrypted_metadata: &str,
        metadata_nonce: &str,
        encrypted_file_key: &str,
        file_key_nonce: &str,
    ) -> Result<String> {
        // Upload-Metadata values are base64 per the tus spec (the metadata
        // strings are themselves already base64 — double-encoded, matching Go).
        let upload_meta = [
            format!("collectionId {}", b64(collection_id.as_bytes())),
            format!("encryptedMetadata {}", b64(encrypted_metadata.as_bytes())),
            format!("metadataNonce {}", b64(metadata_nonce.as_bytes())),
            format!("encryptedFileKey {}", b64(encrypted_file_key.as_bytes())),
            format!("fileKeyNonce {}", b64(file_key_nonce.as_bytes())),
        ]
        .join(",");

        let resp = self
            .upload_request(Method::POST, "/uploads/")
            .header("Tus-Resumable", TUS_VERSION)
            .header("Upload-Length", total_bytes.to_string())
            .header("Upload-Metadata", upload_meta)
            .send()?;

        if resp.status().as_u16() != 201 {
            return Err(api_error(resp)).context("tus create");
        }
        let loc = resp
            .headers()
            .get("Location")
            .and_then(|v| v.to_str().ok())
            .unwrap_or_default()
            .to_string();
        match loc.rsplit('/').next() {
            Some(id) if !id.is_empty() => Ok(id.to_string()),
            _ => bail!("tus create: missing/garbled Location header {loc:?}"),
        }
    }

    /// Returns `(offset, length)` for resume. Mirrors `TusHead`.
    pub fn tus_head(&self, upload_id: &str) -> Result<(i64, i64)> {
        let resp = self
            .request(Method::HEAD, &format!("/uploads/{upload_id}"))
            .header("Tus-Resumable", TUS_VERSION)
            .send()?;
        if resp.status().as_u16() != 200 {
            return Err(api_error(resp)).context("tus head");
        }
        let offset = header_i64(&resp, "Upload-Offset");
        let length = header_i64(&resp, "Upload-Length");
        Ok((offset, length))
    }

    /// Ships one chunk; returns `(new_offset, file_id)` — `file_id` is only set
    /// on the final chunk. Mirrors `TusPatch`.
    pub fn tus_patch(&self, upload_id: &str, offset: i64, body: Vec<u8>) -> Result<(i64, String)> {
        let len = body.len();
        let resp = self
            .upload_request(Method::PATCH, &format!("/uploads/{upload_id}"))
            .header("Tus-Resumable", TUS_VERSION)
            .header("Upload-Offset", offset.to_string())
            .header("Content-Type", "application/offset+octet-stream")
            .header("Content-Length", len.to_string())
            .body(body)
            .send()?;
        if resp.status().as_u16() != 204 {
            return Err(api_error(resp)).context("tus patch");
        }
        let new_offset = header_i64(&resp, "Upload-Offset");
        let file_id = resp
            .headers()
            .get("X-Kutup-File-Id")
            .and_then(|v| v.to_str().ok())
            .unwrap_or_default()
            .to_string();
        Ok((new_offset, file_id))
    }

    /// Cancels an in-flight upload (best-effort). Mirrors `TusDelete`.
    pub fn tus_delete(&self, upload_id: &str) -> Result<()> {
        let resp = self
            .request(Method::DELETE, &format!("/uploads/{upload_id}"))
            .header("Tus-Resumable", TUS_VERSION)
            .send()?;
        let code = resp.status().as_u16();
        if code != 204 && code != 404 {
            return Err(api_error(resp)).context("tus delete");
        }
        Ok(())
    }
}

fn header_i64(resp: &reqwest::blocking::Response, name: &str) -> i64 {
    resp.headers()
        .get(name)
        .and_then(|v| v.to_str().ok())
        .and_then(|s| s.parse().ok())
        .unwrap_or(0)
}
